File: day4.scala
package io.cjf.aoc2024.day4

import scala.io.Source

val inputs = Source.fromFile("./inputs/day4.txt").mkString

enum SearchState:
  case Begin
  case X
  case M
  case A

enum SearchResult:
  case Found
  case NotFound

case class Coord(val row: Int, val col: Int):
  def ul: Coord = Coord(row - 1, col - 1)
  def ur: Coord = Coord(row - 1, col + 1)
  def ll: Coord = Coord(row + 1, col - 1)
  def lr: Coord = Coord(row + 1, col + 1)

class Grid(inp: String):
  val grid = inp.linesIterator.map(_.toCharArray().toSeq).toSeq

  def apply(c: Coord): Option[Char] =
    grid.lift(c.row).flatMap(_.lift(c.col))

  def search_from(
      state: SearchState,
      c: Coord,
      delta: Option[Coord]
  ): Seq[SearchResult] =
    import SearchState._
    val newState = (state, this(c)) match
      case (Begin, Some('X')) => X
      case (X, Some('M'))     => M
      case (M, Some('A'))     => A
      case (A, Some('S'))     => return Seq(SearchResult.Found)
      case _                  => return Seq(SearchResult.NotFound)

    delta match
      case Some(d) =>
        search_from(newState, Coord(c.row + d.row, c.col + d.col), Some(d))
      case None =>
        (-1 until 2).flatMap { row =>
          (-1 until 2).flatMap { col =>
            search_from(
              newState,
              Coord(c.row + row, c.col + col),
              Some(Coord(row, col))
            )
          }
        }

  def has_x_mas_at(c: Coord): Boolean =
    this(c) == Some('A') && (
      (this(c.ul) == Some('M') && this(c.ll) == Some('M') && this(c.ur) == Some(
        'S'
      ) && this(
        c.lr
      ) == Some('S')) ||
        (
          this(c.ul) == Some('M') && this(c.ur) == Some('M') && this(
            c.ll
          ) == Some('S') && this(c.lr) == Some('S')
        ) ||

        (this(c.ur) == Some('M') && this(c.lr) == Some('M') && this(
          c.ul
        ) == Some('S') && this(c.ll) == Some('S')) ||

        (this(c.ll) == Some('M') && this(c.lr) == Some('M') && this(
          c.ul
        ) == Some('S') && this(c.ur) == Some('S'))
    )

  def index_iter: Seq[Coord] =
    for
      row <- 0 until grid.length
      col <- 0 until grid(row).length
    yield Coord(row, col)

def part1(inp: String) =
  val grid = Grid(inp)
  grid.index_iter
    .flatMap { c => grid.search_from(SearchState.Begin, c, None) }
    .count(_ == SearchResult.Found)

def part2(inp: String) =
  val grid = Grid(inp)
  grid.index_iter
    .map(grid.has_x_mas_at)
    .count(c => c)

@main def main() =
  println(s"Part 1: ${part1(inputs)}")
  println(s"Part 2: ${part2(inputs)}")

File: src/bin/day4.rs
use adventofcode2024::coords::{Coord, CoordDiff};

fn inputs() -> String {
    std::fs::read_to_string("./inputs/day4.txt").unwrap()
}

#[derive(Clone, Copy)]
enum SearchState {
    Begin,
    X,
    M,
    A,
}

#[derive(Clone, Copy)]
enum SearchResult {
    Found,
    NotFound,
}

struct Grid {
    grid: Vec<Vec<char>>,
}

impl Grid {
    fn new(inp: &str) -> Grid {
        Grid {
            grid: inp.lines().map(|l| l.chars().collect()).collect(),
        }
    }
    fn get(&self, c: Coord) -> Option<char> {
        let row: usize = c.row.try_into().ok()?;
        let col: usize = c.col.try_into().ok()?;
        self.grid.get(row).and_then(|v| v.get(col).copied())
    }
    fn search_from(
        &self,
        state: SearchState,
        c: Coord,
        delta: Option<CoordDiff>,
    ) -> Vec<SearchResult> {
        use SearchState::*;
        let new_state = match (state, self.get(c)) {
            (Begin, Some('X')) => X,
            (X, Some('M')) => M,
            (M, Some('A')) => A,
            (A, Some('S')) => return vec![SearchResult::Found],
            _ => return vec![SearchResult::NotFound],
        };
        match delta {
            Some(d) => self.search_from(new_state, c + d, Some(d)),
            None => (-1..=1)
                .flat_map(|row| {
                    (-1..=1).flat_map(move |col| {
                        let diff = CoordDiff {
                            rows: row,
                            cols: col,
                        };
                        self.search_from(new_state, c + diff, Some(diff))
                    })
                })
                .collect(),
        }
    }
    fn has_x_mas_at(&self, c: Coord) -> bool {
        self.get(c) == Some('A')
            && (false // silly hack to get rustfmt to make all the conditions indented
                || (self.get(c.ul()) == Some('M')
                    && self.get(c.ll()) == Some('M')
                    && self.get(c.ur()) == Some('S')
                    && self.get(c.lr()) == Some('S'))
                || (self.get(c.ul()) == Some('M')
                    && self.get(c.ur()) == Some('M')
                    && self.get(c.ll()) == Some('S')
                    && self.get(c.lr()) == Some('S'))
                || (self.get(c.ur()) == Some('M')
                    && self.get(c.lr()) == Some('M')
                    && self.get(c.ul()) == Some('S')
                    && self.get(c.ll()) == Some('S'))
                || (self.get(c.ll()) == Some('M')
                    && self.get(c.lr()) == Some('M')
                    && self.get(c.ul()) == Some('S')
                    && self.get(c.ur()) == Some('S')))
    }

    fn index_iter(&self) -> IndexIter<'_> {
        IndexIter {
            parent: self,
            curr_pos: Coord { row: 0, col: 0 },
        }
    }
}

struct IndexIter<'a> {
    parent: &'a Grid,
    curr_pos: Coord,
}

impl Iterator for IndexIter<'_> {
    type Item = Coord;

    fn next(&mut self) -> Option<Self::Item> {
        if self.curr_pos.row >= self.parent.grid.len().try_into().unwrap() {
            None
        } else {
            let to_return = self.curr_pos;
            self.curr_pos.col += 1;
            if self.curr_pos.col
                >= self.parent.grid[TryInto::<usize>::try_into(self.curr_pos.row).unwrap()]
                    .len()
                    .try_into()
                    .unwrap()
            {
                self.curr_pos.col = 0;
                self.curr_pos.row += 1;
            }
            Some(to_return)
        }
    }
}

fn part1(inp: &str) -> usize {
    let grid = Grid::new(inp);
    grid.index_iter()
        .flat_map(|c| grid.search_from(SearchState::Begin, c, None))
        .filter(|it| matches!(it, SearchResult::Found))
        .count()
}

fn part2(inp: &str) -> usize {
    let grid = Grid::new(inp);
    grid.index_iter()
        .map(|c| grid.has_x_mas_at(c))
        .filter(|it| *it)
        .count()
}

fn main() {
    println!("Part 1: {}", part1(&inputs()));
    println!("Part 2: {}", part2(&inputs()));
}
